walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,109 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> attendees = request.getAttendees();
+
+    // If the requested time is longer than the entire day.
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        return Arrays.asList();
+    }
+
+    if (attendees.isEmpty() && !request.getOptionalAttendees().isEmpty()) {
+        attendees = request.getOptionalAttendees();
+    }
+
+    // If there are no attendees or there are no events during the day.
+    if (attendees.isEmpty() || events.isEmpty()) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // Attending list & Optional list
+    List<TimeRange> eventsAL = new ArrayList<>();
+    List<TimeRange> eventsOL = new ArrayList<>();
+
+    // Add events that share attendees.
+    for (Event event : events) {
+        if (!Collections.disjoint(event.getAttendees(), attendees)) {
+            eventsAL.add(event.getWhen());
+        } else if (!Collections.disjoint(event.getAttendees(), request.getOptionalAttendees())) {
+            eventsOL.add(event.getWhen());
+        }
+    }
+
+    // Check if one event contains another.
+    for (int i = 0; i < eventsAL.size() - 1; i++) {
+        if (eventsAL.get(i).contains(eventsAL.get(i + 1))) {
+            eventsAL.remove(eventsAL.get(i + 1));
+        }
+    }
+
+    List<TimeRange> openSpaces = openSlots(eventsAL);
+    Collection<TimeRange> possibilities = new ArrayList<>();
+
+    for (int i = 0; i < openSpaces.size(); i++) {
+
+        // If its only one free space during the day and it's the entire day.
+        if (openSpaces.size() == 1 && openSpaces.get(i).duration() == TimeRange.WHOLE_DAY.duration() - 1) {
+            possibilities.add(TimeRange.fromStartEnd(0, TimeRange.WHOLE_DAY.duration(), false));
+            return possibilities;
+        }
+
+        // If the duration of the free space is longer or equal to the requested duration.
+        if (openSpaces.get(i).duration() >= request.getDuration()) {
+            possibilities.add(TimeRange.fromStartEnd(openSpaces.get(i).start(), openSpaces.get(i).end(), false));
+        }
+    }
+
+    // If possible add the optional attendees.
+    if (possibilities.size() > 1) {
+        for (TimeRange timeAL : possibilities) {
+            for (TimeRange timeOL : eventsOL) {
+                if (timeAL.equals(timeOL)) {
+                    possibilities.remove(timeAL);
+                }
+            }
+        }
+    }
+
+    return possibilities;
+  }
+
+  /**
+   * Get the free slots on the day.
+   */
+  public List<TimeRange> openSlots(List<TimeRange> events) {
+      
```

Comment:
Nit: Remove blank line at start of method.

Reply:
Done!